csrc/id_model/id_model.cpp

```diff
@@ -0,0 +1,292 @@
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <id_model/id_model.h>
+#include <id_model/to_string.h>
+
+#include <device_lower/analysis/trivial_broadcast.h>
+#include <device_lower/lower2device.h>
+#include <device_lower/utils.h>
+#include <disjoint_set.h>
+#include <ir/utils.h>
+#include <root_domain_map.h>
+#include <transform_iter.h>
+
+#include <tuple>
+#include <typeinfo>
+#include <utility>
+
+namespace nvfuser {
+
+namespace {
+
+// Map through loop swizzles, as input/output IterDomains are exact, only the
+// order they're traversed differs.
+void mapThroughLoopSwizzles(ValGraph& graph) {
+  std::vector<Swizzle2D*> all_swizzles;
+
+  for (const auto& expr_set :
+       std::as_const(graph).disjointExprSets().disjointSets()) {
+    auto swizzles_in_expr_set = ir_utils::filterByType<Swizzle2D>(
+        expr_set->vector().begin(), expr_set->vector().end());
+    all_swizzles.insert(
+        all_swizzles.end(),
+        swizzles_in_expr_set.begin(),
+        swizzles_in_expr_set.end());
+  }
+
+  for (auto swizzle : all_swizzles) {
+    if (swizzle->swizzleMode() == SwizzleMode::Loop) {
+      graph.mapVals(swizzle->inX(), swizzle->outX());
+      graph.mapVals(swizzle->inY(), swizzle->outY());
+    }
+  }
+}
+
+} // namespace
+
+IdModel::IdModel(
+    const std::vector<Expr*>& exprs,
+    const std::vector<TensorView*>& additional_tvs,
+    bool allow_self_mapping) {
+  build(exprs, additional_tvs);
+}
+
+IdModel::IdModel(const std::vector<Expr*>& exprs, bool allow_self_mapping)
+    : IdModel(exprs, {}, allow_self_mapping) {}
+
+IdModel::IdModel(Fusion* fusion, bool allow_self_mapping) {
+  std::vector<TensorView*> inputs_and_outputs;
+  {
+    auto inp_tvs = ir_utils::filterByType<TensorView>(fusion->inputs());
+    inputs_and_outputs.insert(
+        inputs_and_outputs.begin(), inp_tvs.begin(), inp_tvs.end());
+  }
+  {
+    auto out_tvs = ir_utils::filterByType<TensorView>(fusion->outputs());
+    inputs_and_outputs.insert(
+        inputs_and_outputs.begin(), out_tvs.begin(), out_tvs.end());
+  }
+
+  build(fusion->exprs(), inputs_and_outputs);
+}
+
+const ValGraph& IdModel::idGraph(IdMappingMode mode) const {
+  auto graph_it = id_graphs_.find(mode);
+  NVF_ERROR(graph_it != id_graphs_.end());
+  return graph_it->second;
+}
+
+ValGraph& IdModel::idGraph(IdMappingMode mode) {
+  auto graph_it = id_graphs_.find(mode);
+  NVF_ERROR(graph_it != id_graphs_.end());
+  return graph_it->second;
+}
+
+Expr* IdModel::idUse(IterDomain* id) const {
+  auto use_it = id_uses_.find(id);
+  if (use_it == id_uses_.end()) {
+    return nullptr;
+  }
+  return use_it->second.front();
+}
+
+Expr* IdModel::idDef(IterDomain* id) const {
+  auto def_it = id_definitions_.find(id);
+  if (def_it == id_definitions_.end()) {
+    return nullptr;
+  }
+  return def_it->second.front();
+}
+
+void IdModel::buildIterDomainDefinitionsAndUses(
```

Comment:
The logic here reminds me of `DependencyCheck::getAllExprsBetween`. How is it different from
```
for (each TensorView in all_tvs) {
 for (each Expr in DependencyCheck::getAllExprsBetween(tv->getRootDomain(), tv->getMaybeRFactorDomain())) {
   add Expr's def-use to id_defs_ and id_uses_;
 }
} 
```
?

Reply:
Yes, it's actually quite similar since `allIDsOf` uses `getAllValsBetween`. Note that we also look at domains outside the root-rfactor path.